core/src/main/java/org/apache/accumulo/core/summary/Gatherer.java

```diff
@@ -0,0 +1,513 @@
+        for (Map<String,List<TRowRange>> files : partition(allFiles, 500)) {
+          if (pfiles.failedFiles.size() > 0) {
+            // there was a previous failure on this tserver, so just fail the rest of the files
+            pfiles.failedFiles.addAll(files.keySet());
+            continue;
+          }
+
+          TSummaries tSums = null;
+          try {
+            tSums = client.getSummariesFromFiles(tinfo, ctx.rpcCreds(), getRequest(), files);
+          } catch (TApplicationException tae) {
+            throw tae;
+          } catch (TTransportException e) {
+            pfiles.failedFiles.addAll(files.keySet());
+            continue;
+          }
+
+          pfiles.summaries.merge(new SummaryCollection(tSums), factory);
+        }
+      } finally {
+        ThriftUtil.returnClient(client);
+      }
+
+      return pfiles;
+    }
+  }
+
+  /**
+   * This methods reads a subset of file paths into memory and groups them by location. Then it request sumaries for files from each location/tablet server.
+   */
+  public SummaryCollection processPartition(int modulus, int remainder) throws TableNotFoundException, AccumuloException, AccumuloSecurityException {
+
+    Predicate<String> hashingFileSelector = file -> Math.abs(Hashing.murmur3_32().hashString(file).asInt()) % modulus == remainder;
+
+    Map<String,Map<String,List<TRowRange>>> filesGBL = getFilesGroupedByLocation(hashingFileSelector);
+
+    SummaryCollection partitionSummaries = new SummaryCollection();
+
+    ExecutorService execSrv = Executors.newCachedThreadPool();
+    try {
+      TInfo tinfo = Tracer.traceInfo();
+      while (true) {
+        Set<String> failedFiles = new HashSet<>();
+
+        List<Future<ProcessedFiles>> futures = new ArrayList<>();
+        for (Entry<String,Map<String,List<TRowRange>>> entry : filesGBL.entrySet()) {
+          HostAndPort location = HostAndPort.fromString(entry.getKey());
+          Map<String,List<TRowRange>> allFiles = entry.getValue();
+
+          Future<ProcessedFiles> future = execSrv.submit(new FilesProcessor(tinfo, location, allFiles));
+          futures.add(future);
+        }
+
+        for (Future<ProcessedFiles> future : futures) {
+          try {
+            ProcessedFiles pfiles = future.get();
+            if (pfiles.summaries.getTotalFiles() > 0) {
+              partitionSummaries.merge(pfiles.summaries, factory);
+            }
+
+            failedFiles.addAll(pfiles.failedFiles);
+          } catch (InterruptedException | ExecutionException e) {
+            throw new AccumuloException(e);
+          }
+        }
+
+        if (failedFiles.size() > 0) {
+          // get new locations just for failed files
+          Predicate<String> fileSelector = hashingFileSelector.and(file -> failedFiles.contains(file));
+          filesGBL = getFilesGroupedByLocation(fileSelector);
+          UtilWaitThread.sleep(250);
+        } else {
+          break;
```

Comment:
Just make this `return partitionSummaries;`? Not sure if I'm missing an execution branch.